pkg/sql/schemachanger/scplan/plan.go

```diff
@@ -72,32 +70,34 @@ func MakePlan(initial scpb.State, params Params) (p Plan, err error) {
 		}
 	}()
 
-	p = Plan{
-		Initial: initial,
-		Params:  params,
+	p.Graph = buildGraph(initial)
+	p.Stages = scstage.BuildStages(initial, params.ExecutionPhase, p.Graph, params.SchemaChangerJobIDSupplier)
+	if n := len(p.Stages); n > 0 && p.Stages[n-1].Phase > scop.PreCommitPhase {
+		// Only get the job ID if it's actually been assigned already.
+		p.JobID = params.SchemaChangerJobIDSupplier()
 	}
+	if err = scstage.ValidateStages(p.Stages, p.Graph); err != nil {
+		panic(errors.Wrapf(err, "invalid execution plan"))
+	}
+	return p, nil
+}
+
+func buildGraph(initial scpb.State) *scgraph.Graph {
 	g, err := opgen.BuildGraph(initial)
 	if err != nil {
-		return p, err
-	}
-	p.Graph = g
-	if err := deprules.Apply(g); err != nil {
-		return p, err
+		panic(errors.Wrapf(err, "build graph op edges"))
 	}
-	optimizedGraph, err := scopt.OptimizePlan(g)
+	err = deprules.Apply(g)
 	if err != nil {
-		return p, err
+		panic(errors.Wrapf(err, "build graph dep edges"))
 	}
-	p.Graph = optimizedGraph
-
-	p.stages = scstage.BuildStages(initial, params.ExecutionPhase, optimizedGraph)
-	if err = scstage.ValidateStages(p.stages); err != nil {
-		return p, errors.WithAssertionFailure(errors.Wrapf(err, "invalid basic execution plan"))
+	err = g.Validate()
+	if err != nil {
+		panic(errors.Wrapf(err, "validate graph"))
 	}
-	p.stages = scstage.CollapseStages(p.stages)
-	p.stages, p.JobID = scstage.AugmentStagesForJob(p.stages, params.ExecutionPhase, params.JobIDGenerator)
-	if err = scstage.ValidateStages(p.stages); err != nil {
-		return p, errors.WithAssertionFailure(errors.Wrapf(err, "invalid improved execution plan"))
+	g, err = scopt.OptimizeGraph(g)
+	if err != nil {
+		panic(errors.Wrapf(err, "mark op edges as no-op"))
 	}
-	return p, nil
+	return g
```

Comment:
It upsets me that we don't validate at this point. I suspect it upsets you too. The reasoning is clear, I don't think that the resultant graph would be valid. You can see this in the plan tests where we just now leave the columns in a non-terminal state when dropping the table. I'm fine deferring these changes, but at this point, this call is the scariest part of this project. The builder might have implementation issues, but they don't upset me philosophically like this part does.